File: tools/fetch-vitamins.php
<?php
/**
 * Converts CSV files downloaded from https://companion.pokeclicker.com into
 * JSON files containing only the necessary data for the user script to
 * optimally distribute all the vitamins.
 *
 * As of now, the source files need to be placed with the script. It would be
 * nice, if they could be downloaded on-demand at some point but, because of
 * their dynamic nature, unfortunately, there is no straightforward way to get
 * a download link for them.
 */

declare(strict_types=1);

enum Region: int
{
    case Kanto = 0;
    case Johto = 1;
    case Hoenn = 2;
    case Sinnoh = 3;
    case Unova = 4;
    case Kalos = 5;
    case Alola = 6;
    case Galar = 7;
}

enum Vitamin: int
{
    case Protein = 0;
    case Calcium = 1;
    case Carbos = 2;
}

$inputFiles = [
    Region::Kanto->value => __DIR__ . '/data/01-kanto.csv',
    Region::Johto->value => __DIR__ . '/data/02-johto.csv',
    Region::Hoenn->value => __DIR__ . '/data/03-hoenn.csv',
    Region::Sinnoh->value => __DIR__ . '/data/04-sinnoh.csv',
    Region::Unova->value => __DIR__ . '/data/05-unova.csv',
    Region::Kalos->value => __DIR__ . '/data/06-kalos.csv',
    Region::Alola->value => __DIR__ . '/data/07-alola.csv',
    Region::Galar->value => __DIR__ . '/data/08-galar.csv',
];

$regions = [
    Region::Kanto->value => [],
    Region::Johto->value => [],
    Region::Hoenn->value => [],
    Region::Sinnoh->value => [],
    Region::Unova->value => [],
    Region::Kalos->value => [],
    Region::Alola->value => [],
    Region::Galar->value => [],
];

foreach ($inputFiles as $regionId => $inputFileName) {
    $currentRow = 0;
    $firstRow = [];
    $fd = fopen($inputFileName, 'r');

    while ($row = fgetcsv($fd)) {
        if ($currentRow++ === 0) {
            $firstRow = $row;

            continue;
        }

        $parsedRow = array_combine($firstRow, $row); // re-index from header values
        $regions[$regionId][$parsedRow['#']] = [
            Vitamin::Protein->value => (int)$parsedRow['Optimal Protein'],
            Vitamin::Calcium->value => (int)$parsedRow['Optimal Calcium'],
            Vitamin::Carbos->value => (int)$parsedRow['Optimal Carbos'],
        ];
    }

    fclose($fd);

    $targetFilename = dirname($inputFileName) . '/' . basename($inputFileName, '.csv') . '.json';
    file_put_contents($targetFilename, json_encode($regions[$regionId]));
}

File: tools/fetch-vitamins.py
import csv
import json
from enum import IntEnum
from pathlib import Path

class Region(IntEnum):
    Kanto = 0
    Johto = 1
    Hoenn = 2
    Sinnoh = 3
    Unova = 4
    Kalos = 5
    Alola = 6
    Galar = 7

class Vitamin(IntEnum):
    Protein = 0
    Calcium = 1
    Carbos = 2

data_dir = Path(__file__).parent / 'data'
input_files = {
    Region.Kanto: data_dir / '01-kanto.csv',
    Region.Johto: data_dir / '02-johto.csv',
    Region.Hoenn: data_dir / '03-hoenn.csv',
    Region.Sinnoh: data_dir / '04-sinnoh.csv',
    Region.Unova: data_dir / '05-unova.csv',
    Region.Kalos: data_dir / '06-kalos.csv',
    Region.Alola: data_dir / '07-alola.csv',
    Region.Galar: data_dir / '08-galar.csv',
}

regions = {region: {} for region in Region}

for region, input_file in input_files.items():
    with input_file.open('r', newline='', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            regions[region][row['#']] = [
                int(row['Optimal Protein']),
                int(row['Optimal Calcium']),
                int(row['Optimal Carbos'])
            ]

    output_file = input_file.with_suffix('.json')
    with output_file.open('w', encoding='utf-8') as jsonfile:
        json.dump(regions[region], jsonfile, separators=(',', ':'), ensure_ascii=False)